libr/bin/dwarf.c

```diff
@@ -1319,321 +1519,523 @@ static void r_bin_dwarf_dump_debug_info(FILE *f, const RBinDwarfDebugInfo *inf)
+
 	const ut8 *buf = obuf;
 	const ut8 *buf_end = obuf + obuf_len;
 	size_t j;
 
-	if (!spec || !value || !hdr || !obuf || obuf_len < 1) {
-		return NULL;
-	}
+	r_return_val_if_fail(def && value && hdr && obuf && obuf_len >= 1, NULL);
 
-	value->form = spec->attr_form;
-	value->name = spec->attr_name;
-	value->encoding.block.data = NULL;
-	value->encoding.str_struct.string = NULL;
-	value->encoding.str_struct.offset = 0;
+	value->attr_form = def->attr_form;
+	value->attr_name = def->attr_name;
+	value->block.data = NULL;
+	value->string.content = NULL;
+	value->string.offset = 0;
 
-	switch (spec->attr_form) {
+	switch (def->attr_form) {
 	case DW_FORM_addr:
-		switch (hdr->pointer_size) {
+		switch (hdr->address_size) {
 		case 1:
-			value->encoding.address = READ8 (buf);
+			value->address = READ8 (buf);
 			break;
 		case 2:
-			value->encoding.address = READ16 (buf);
+			value->address = READ16 (buf);
 			break;
 		case 4:
-			value->encoding.address = READ32 (buf);
+			value->address = READ32 (buf);
 			break;
 		case 8:
-			value->encoding.address = READ64 (buf);
+			value->address = READ64 (buf);
 			break;
 		default:
-			eprintf ("DWARF: Unexpected pointer size: %u\n", (unsigned)hdr->pointer_size);
+			eprintf ("DWARF: Unexpected pointer size: %u\n", (unsigned)hdr->address_size);
 			return NULL;
 		}
 		break;
-	case DW_FORM_block2:
-		value->encoding.block.length = READ16 (buf);
-		if (value->encoding.block.length > 0) {
-			value->encoding.block.data = calloc (sizeof(ut8), value->encoding.block.length);
-			for (j = 0; j < value->encoding.block.length; j++) {
-				value->encoding.block.data[j] = READ (buf, ut8);
-			}
-		}
-		break;
-	case DW_FORM_block4:
-		value->encoding.block.length = READ32 (buf);
-		if (value->encoding.block.length > 0) {
-			ut8 *data = calloc (sizeof (ut8), value->encoding.block.length);
-			if (data) {
-				for (j = 0; j < value->encoding.block.length; j++) {
-					data[j] = READ (buf, ut8);
-				}
-			}
-			value->encoding.block.data = data;
-		}
+	case DW_FORM_data1:
+		value->data = READ8 (buf);
 		break;
 	case DW_FORM_data2:
-		value->encoding.data = READ16 (buf);
+		value->data = READ16 (buf);
 		break;
 	case DW_FORM_data4:
-		value->encoding.data = READ32 (buf);
+		value->data = READ32 (buf);
 		break;
 	case DW_FORM_data8:
-		value->encoding.data = READ64 (buf);
+		value->data = READ64 (buf);
+		break;
+	case DW_FORM_data16: // TODO Fix this, right now I just read the data, but I need to make storage for it
+		value->data = READ64 (buf);
+		value->data = READ64 (buf);
+		break;
+	case DW_FORM_sdata:
+		buf = r_leb128 (buf, buf_end - buf, &value->sdata);
+		break;
+	case DW_FORM_udata:
+		buf = r_uleb128 (buf, buf_end - buf, &value->data);
 		break;
 	case DW_FORM_string:
-		value->encoding.str_struct.string = *buf? strdup ((const char*)buf) : NULL;
-		buf += (strlen ((const char*)buf) + 1);
+		value->string.content = *buf ? strdup ((const char *)buf) : NULL;
```

Comment:
ok... i see no check even here, but let's not worry about those for now.. There are probably anyway many more places where these kind of checks are missing :(

Reply:
Yeah, it's pretty hard to make parsing bulletproof. I am not sure how would I check except linear search till the end of buffer for every string (in case of correctly formed dwarf this wouldn't really be much of an overhead)